src/cargo/sources/registry/mod.rs

```diff
@@ -701,51 +701,59 @@ impl<'cfg> RegistrySource<'cfg> {
 }
 
 impl<'cfg> Source for RegistrySource<'cfg> {
-    fn query(&mut self, dep: &Dependency, f: &mut dyn FnMut(Summary)) -> Poll<CargoResult<()>> {
-        // If this is a precise dependency, then it came from a lock file and in
-        // theory the registry is known to contain this version. If, however, we
-        // come back with no summaries, then our registry may need to be
-        // updated, so we fall back to performing a lazy update.
-        if dep.source_id().precise().is_some() && !self.ops.is_updated() {
-            debug!("attempting query without update");
-            let mut called = false;
-            let pend =
-                self.index
-                    .query_inner(dep, &mut *self.ops, &self.yanked_whitelist, &mut |s| {
-                        if dep.matches(&s) {
-                            called = true;
-                            f(s);
-                        }
-                    })?;
-            if pend.is_pending() {
-                return Poll::Pending;
-            }
-            if called {
-                return Poll::Ready(Ok(()));
-            } else {
-                debug!("falling back to an update");
-                self.invalidate_cache();
-                return Poll::Pending;
+    fn query(
+        &mut self,
+        dep: &Dependency,
+        kind: QueryKind,
+        f: &mut dyn FnMut(Summary),
+    ) -> Poll<CargoResult<()>> {
+        match kind {
+            QueryKind::Exact => {
+                // If this is a precise dependency, then it came from a lock file and in
+                // theory the registry is known to contain this version. If, however, we
+                // come back with no summaries, then our registry may need to be
+                // updated, so we fall back to performing a lazy update.
+                if dep.source_id().precise().is_some() && !self.ops.is_updated() {
+                    debug!("attempting query without update");
+                    let mut called = false;
+                    let pend = self.index.query_inner(
+                        dep,
+                        &mut *self.ops,
+                        &self.yanked_whitelist,
+                        &mut |s| {
+                            if dep.matches(&s) {
+                                called = true;
+                                f(s);
+                            }
+                        },
+                    )?;
+                    if pend.is_pending() {
+                        return Poll::Pending;
+                    }
+                    if called {
+                        return Poll::Ready(Ok(()));
+                    } else {
+                        debug!("falling back to an update");
+                        self.invalidate_cache();
+                        return Poll::Pending;
+                    }
+                }
             }
+            QueryKind::Fuzzy => {}
```

Comment:
Was something supposed go here? If not can we have a comment about why?

Reply:
The comment is on the other branch.  If you have thoughts for improving this, I'm open.

I could have just done `if kind == Exact` but I generally prefer `match` with enums in case future cases might need to cause an assumption to be re-evaluated.

I could swap the comment so it goes here but I feel like the other branch needs the description as well.  maybe if I could it to be before the match?

Reply:
There's a page of code between the explanation and this. I did not realize that the comment applied here. Moving the comment outside the match would technically help, but would also make more lines between.

I think semantically the check is part of the if statement. Only `if kind == Exact && dep.source_id().precise().is_some() && !self.ops.is_updated() {` are we in the no-update lockfile situation.

Reply:
Went ahead and id that.